build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/ThirdPartyAuditTask.java

```diff
@@ -355,6 +359,33 @@ private String runForbiddenAPIsCli() throws IOException {
         return forbiddenApisOutput;
     }
 
+    void maybeAddIncubatorModule(JavaExecSpec spec) {
+        if (javaHome.isPresent() == false) {
+            return;
+        }
+        Path releaseFile = Path.of(javaHome.get() + "/release");
+        if (Files.notExists(releaseFile)) {
+            throw new IllegalStateException("JDK missing release file, for: " + javaHome.get());
+        }
+        try {
+            Runtime.Version version = Files.readAllLines(releaseFile).stream()
```

Comment:
@mark-vieira @breskeby ping'ing you here in case you have opinions on this this could be more Gradle idiomatic. 

This is probably the first time we've needed to do this - to support a new Lucene update that is now using the jdk.incubator.vector API.  Note: the release file is always at the top of the JDK distribution, and has a defined format which includes the JAVA_VERSION

Reply:
I don't think we need to do all this, we already have `BuildParams.getRuntimeJavaVersion()` to detect the version of the JAVA_HOME used at runtime.

Reply:
Chris, presumably we need to do something at runtime as well? Does Lucene 9.7 assume these modules will be available at runtime when running against Java 20+? I assume they are doing some multi-release JAR action again? If so, do we need to pass this `--add-modules` argument to ES at runtime when using a Java 20+ JDK?

Reply:
This would indeed be much cleaner, but it doesn't work (or maybe I've not done it right).

```diff
diff --git a/build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/ThirdPartyAuditTask.java b/build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/ThirdPartyAuditTask.java
index bc11fb976104..a3892be9e19d 100644
--- a/build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/ThirdPartyAuditTask.java
+++ b/build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/ThirdPartyAuditTask.java
@@ -337,7 +337,10 @@ public abstract class ThirdPartyAuditTask extends DefaultTask {
             }
             spec.classpath(getForbiddenAPIsClasspath(), classpath);
             // Enable explicitly for each release as appropriate. Just JDK 20 for now, and just the vector module.
-            if (VERSION_20.equals(BuildParams.getRuntimeJavaVersion())) {
+            System.out.println("javaHome.get()= "+ javaHome.get());
+            System.out.println("BuildParams.getRuntimeJavaVersion()= "+ BuildParams.getRuntimeJavaVersion());
+            System.out.println("BuildParams.getRuntimeJavaHome()= "+ BuildParams.getRuntimeJavaHome());
+            if (VERSION_20.getMajorVersion().equals(BuildParams.getRuntimeJavaVersion().getMajorVersion())) {
                 spec.jvmArgs("--add-modules", "jdk.incubator.vector");
             }
             spec.jvmArgs("-Xmx1g");

```

Output:
```
javaHome.get()= /Users/chegar/.gradle/caches/transforms-3/da05774a19e6937af1b3b3d0be6c309f/transformed/macos-20.0.1-aarch64.tar.gz/Contents/Home
BuildParams.getRuntimeJavaVersion()= 18
BuildParams.getRuntimeJavaHome()= /Users/chegar/binaries/jdk-18.0.2.1.jdk/Contents/Home
```

Reply:
> Chris, presumably we need to do something at _actual_ runtime as well? Does Lucene 9.7 assume these modules will be available at runtime when running against Java 20+? I assume they are doing some multi-release JAR action again? If so, do we need to pass this `--add-modules` argument to ES at runtime when using a Java 20+ JDK?

Correct. There is a separate issue that will do this, see https://github.com/elastic/elasticsearch/issues/96370.  We want to upgrade first, then enable separately.

Reply:
This kinda explains why runtime java how is not working (based on my changes above).  https://github.com/elastic/elasticsearch/blob/main/build-tools-internal/src/main/groovy/elasticsearch.runtime-jdk-provision.gradle#L64-L70

There's something funky going on here for third party audit task.

Reply:
Yeah, there's a few things going on here. First, is that this task used the _runtime_ Java version, which can be overridden by the `RUNTIME_JAVA_HOME` property. We do this in our JDK testing as we have to maintain compatbility with back to Java 17 as well as testing EA versions (we'll start testing on Java 21) soon.

https://github.com/elastic/elasticsearch/blob/527cf1d3f552f66b5062b0dea6fcf8819c195018/build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/ThirdPartyAuditPrecommitPlugin.java#L65

When the runtime Java home _isn't_ overriden, we use the bundled JDK. That's done here:

https://github.com/elastic/elasticsearch/blob/4e2f852c08565b02ff49a646373411e6f2926c03/build-tools-internal/src/main/groovy/elasticsearch.runtime-jdk-provision.gradle#L68

So we need some more conditional logic here. Basically, if `isRuntimeJavaHomeSet() == false` then yes, we can assume it's the bundled JDK major version. If it's `true`, then we need to inspect `BuildParams.getRuntimeJavaVesion()`.

Reply:
👍 done [085d38d](https://github.com/elastic/elasticsearch/pull/96433/commits/085d38d24388729ca5e624da54a079e6b18bf3a7)